lib/vrl/stdlib/src/for_each.rs

```diff
@@ -0,0 +1,189 @@
+                required: true,
+            },
+            Parameter {
+                keyword: "recursive",
+                kind: kind::BOOLEAN,
+                required: false,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "iterate object",
+                source: r#"count = 0; for_each({ "a": 1, "b": 2 }) -> |_key, value| { count = count + value }; count"#,
+                result: Ok("3"),
+            },
+            Example {
+                title: "recursively iterate object",
+                source: r#"count = 0; for_each({ "a": 1, "b": { "c": 2, "d": 3 } }, recursive: true) -> |_, value| { count = count + (int(value) ?? 0) }; count"#,
+                result: Ok("6"),
+            },
+            Example {
+                title: "iterate array",
+                source: r#"count = 0; for_each([1,2,3]) -> |index, value| { count = count + index + value }; count"#,
+                result: Ok("9"),
+            },
+            Example {
+                title: "recursively iterate array",
+                source: r#"count = 0; for_each([1,2,[3,4]], recursive: true) -> |index, value| { count = count + index + (int(value) ?? 0) }; count"#,
+                result: Ok("14"),
+            },
+        ]
+    }
+
+    fn compile(
+        &self,
+        _state: (&mut state::LocalEnv, &mut state::ExternalEnv),
+        _ctx: &mut FunctionCompileContext,
+        mut arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+        let recursive = arguments.optional("recursive");
+        let closure = arguments.required_closure()?;
+
+        Ok(Box::new(ForEachFn {
+            value,
+            closure,
+            recursive,
+        }))
+    }
+
+    fn closure(&self) -> Option<closure::Definition> {
+        use closure::{Definition, Input, Output, Variable, VariableKind};
+
+        Some(Definition {
+            inputs: vec![Input {
+                parameter_keyword: "value",
+                kind: Kind::object(Collection::any()).or_array(Collection::any()),
+                variables: vec![
+                    Variable {
+                        kind: VariableKind::TargetInnerKey,
+                    },
+                    Variable {
+                        kind: VariableKind::TargetInnerValue,
+                    },
+                ],
+                output: Output::Kind(Kind::any()),
+                example: Example {
+                    title: "iterate array",
+                    source: r#"for_each([1, 2]) -> |index, value| { .foo = to_int!(.foo) + index + value }"#,
+                    result: Ok("null"),
+                },
+            }],
+            is_iterator: true,
+        })
+    }
+
+    fn call_by_vm(&self, _ctx: &mut Context, _args: &mut VmArgumentList) -> Result<Value> {
+        // TODO: this work will happen in a follow-up PR
+        Ok(Value::Null)
```

Comment:
Same comment as in https://github.com/vectordotdev/vector/pull/12388#discussion_r858430213:

Would error here to give feedback rather than continue silently, in the rare case this is hit

Reply:
Yeah, we can do that (I _think_). I initially added this to avoid the test harness erroring, but I've since added a hardcoded list of tests to temporarily ignore there, so I think we can return an error here insead.

Reply:
Resolved in [9dccfcf](https://github.com/vectordotdev/vector/pull/12382/commits/9dccfcf8e76108e5cf4dbf279e148b6e33aa22f8).